crates/iota-core/src/authority/scorer.rs

```diff
@@ -77,37 +143,315 @@ impl Scorer {
+        .zip(parameters.allowances.iter_major_misbehaviors())
+        .zip(parameters.maximums.iter_major_misbehaviors());
+
+    // Calculate individual metric scores
+    let major_metric_scores = median_major_reports_and_parameters
+        .map(
+            |((median_report_for_a_single_metric, metric_allowance), metric_maximum)| {
+                median_report_single_metric_to_score(
+                    median_report_for_a_single_metric,
+                    *metric_allowance,
+                    *metric_maximum,
+                    1,
+                )
+            },
+        )
+        .collect::<Vec<Vec<u64>>>();
+
+    metrics_scores_to_final_scores(
+        minor_metric_scores,
+        major_metric_scores,
+        parameters.weights,
+        baseline_score,
+        parameters.scale_factor,
+        parameters.max_score,
+    )
+}
+
+fn metrics_scores_to_final_scores(
+    minor_metric_scores: Vec<Vec<u64>>,
+    major_metric_scores: Vec<Vec<u64>>,
+    weights: MisbehaviorsV1<u64>,
+    baseline_score: u64,
+    scale_factor: u64,
+    max_score: u64,
+) -> Vec<u64> {
+    // Initialise the final scores with the baseline score whose value is between 0
+    // and max_score * scale_factor.
+    let committee_size = minor_metric_scores.first().unwrap().len();
+    let mut final_scores = vec![baseline_score * max_score; committee_size];
+    // First, calculate the weights sum of minor misbehavior scores vector. The
+    // values in final_scores will still be between 0 and max_score * scale_factor
+    minor_metric_scores
+        .iter()
+        .zip(weights.iter_minor_misbehaviors())
+        .for_each(|(scores, weight)| {
+            for (i, &score) in scores.iter().enumerate() {
+                final_scores[i] += score * weight;
+            }
+        });
+    // Then, multiply by each major misbehavior score which is a value of either 0
+    // or 1.
+    major_metric_scores.iter().for_each(|scores| {
+        for (i, &score) in scores.iter().enumerate() {
+            final_scores[i] *= score;
+        }
+    });
+    // Finally, divide by the scale factor and scale to max_score
+    for score in final_scores.iter_mut() {
+        *score /= scale_factor;
+    }
+    final_scores
+}
+
+// Calculate the metric scores for a single metric's median report vector. It
+// returns a vector of values between 0 and the max score for that metric.
+fn median_report_single_metric_to_score(
+    median_report_for_metric: &MedianMetricVec,
+    metric_allowance: u64,
+    metric_max: u64,
+    max_metric_score: u64,
+) -> Vec<u64> {
+    median_report_for_metric
+        .iter()
+        .map(|&report| metric_to_score(report, metric_allowance, metric_max, max_metric_score))
+        .collect()
+}
+
+fn metric_to_score(value: u64, allowance: u64, max: u64, max_score: u64) -> u64 {
+    if value <= allowance {
+        max_score
+    } else if value >= max {
+        0
+    } else {
+        // Add overflow checks
```

Comment:
```suggestion
        // TODO: add overflow checks
```
Also it might be good to define a corresponding issue for the TODO if you haven't PRed and/or planned.

Reply:
Done, issue created https://github.com/iotaledger/iota-private/issues/242